File: MCD_Python/lib/core/detector.py
"""
Core detector class for MTG card detector.
"""

import glob
import os
import pickle
import cv2

from lib.models import ReferenceImage, TestImage
from lib.image import segment_image
from lib.core.recognition import phash_compare


class MagicCardDetector:
    """
    MTG card detector class.
    """
    def __init__(self, output_path=None):
        """
        Initialize the detector.
        output_path: Optional path for saving results.
        """
        self.output_path = output_path
        self.reference_images = []
        self.test_images = []

        self.verbose = False
        self.visual = False

        self.hash_separation_thr = 4.0
        self.thr_lvl = 70

        self.clahe = cv2.createCLAHE(clipLimit=2.0,
                                    tileGridSize=(8, 8))

    def export_reference_data(self, path):
        """
        Exports the phash and card name of the reference data list.
        """
        hlist = []
        for image in self.reference_images:
            hlist.append(ReferenceImage(image.name,
                                       None,
                                       None,
                                       image.phash))

        with open(path, 'wb') as fhandle:
            pickle.dump(hlist, fhandle)

    def read_prehashed_reference_data(self, path):
        """
        Reads pre-calculated hashes of the reference images.
        """
        print('Reading prehashed data from ' + str(path))
        print('...', end=' ')
        with open(path, 'rb') as filename:
            hashed_list = pickle.load(filename)
        for ref_im in hashed_list:
            self.reference_images.append(
                ReferenceImage(ref_im.name, None, self.clahe, ref_im.phash))
        print('Done.')

    def read_and_adjust_reference_images(self, path):
        """
        Reads and histogram-adjusts the reference image set.
        Pre-calculates the hashes of the images.
        """
        print('Reading images from ' + str(path))
        print('...', end=' ')
        filenames = glob.glob(path + '*.jpg')
        for filename in filenames:
            img = cv2.imread(filename)
            img_name = filename.split(path)[1]
            self.reference_images.append(
                ReferenceImage(img_name, img, self.clahe))
        print('Done.')

    def read_and_adjust_test_images(self, path):
        """
        Reads and histogram-adjusts the test image set.
        """
        maxsize = 1000
        print('Reading images from ' + str(path))
        print('...', end=' ')
        filenames = glob.glob(path.rstrip('/') + '/*.jpg')
        for filename in filenames:
            img = cv2.imread(filename)
            if min(img.shape[0], img.shape[1]) > maxsize:
                scalef = maxsize / min(img.shape[0], img.shape[1])
                img = cv2.resize(img,
                                (int(img.shape[1] * scalef),
                                 int(img.shape[0] * scalef)),
                                interpolation=cv2.INTER_AREA)

            img_name = os.path.basename(filename)
            self.test_images.append(
                TestImage(img_name, img, self.clahe))
        print('Done.')

    def recognize_segment(self, image_segment):
        """
        Wrapper for different segmented image recognition algorithms.
        """
        return phash_compare(image_segment, self.reference_images, 
                           self.hash_separation_thr, self.verbose)

    def recognize_cards_in_image(self, test_image, contouring_mode):
        """
        Tries to recognize cards from the image specified.
        The image has been read in and adjusted previously,
        and is contained in the internal data list of the class.
        """
        print('Segmentating card candidates out of the image...')
        print('Using ' + str(contouring_mode) + ' algorithm.')

        test_image.candidate_list.clear()
        segment_image(test_image, contouring_mode=contouring_mode, 
                    visual=self.visual, verbose=self.verbose)

        print('Done. Found ' +
              str(len(test_image.candidate_list)) + ' candidates.')
        print('Recognizing candidates.')

        for i_cand, candidate in enumerate(test_image.candidate_list):
            im_seg = candidate.image
            if self.verbose:
                print(str(i_cand + 1) + " / " +
                      str(len(test_image.candidate_list)))

            # Easy fragment / duplicate detection
            for other_candidate in test_image.candidate_list:
                if (other_candidate.is_recognized and
                        not other_candidate.is_fragment):
                    if other_candidate.contains(candidate):
                        candidate.is_fragment = True
            if not candidate.is_fragment:
                (candidate.is_recognized,
                 candidate.recognition_score,
                 candidate.name) = self.recognize_segment(im_seg)

        print('Done. Found ' +
              str(len(test_image.return_recognized())) +
              ' cards.')
        if self.verbose:
            for card in test_image.return_recognized():
                print(card.name + '; S = ' + str(card.recognition_score))
        print('Removing duplicates...')
        # Final fragment detection
        test_image.mark_fragments()
        print('Done.')

    def run_recognition(self, image_index=None):
        """
        The top-level image recognition method.
        Wrapper for switching to different algorithms and re-trying.
        """
        if image_index is None:
            image_index = range(len(self.test_images))
        elif not isinstance(image_index, list):
            image_index = [image_index]
        for i in image_index:
            test_image = self.test_images[i]
            print('Accessing image ' + test_image.name)

            if self.visual:
                import matplotlib.pyplot as plt
                print('Original image')
                plt.imshow(cv2.cvtColor(test_image.original,
                                       cv2.COLOR_BGR2RGB))
                plt.show()

            alg_list = ['adaptive', 'rgb']

            for alg in alg_list:
                self.recognize_cards_in_image(test_image, alg)
                test_image.discard_unrecognized_candidates()
                if (not test_image.may_contain_more_cards() or
                        len(test_image.return_recognized()) > 5):
                    break

            print('Plotting and saving the results...')
            test_image.plot_image_with_recognized(self.output_path, self.visual)
            print('Done.')
            test_image.print_recognized()
        print('Recognition done.')

File: mcd_dart/lib/core/detector.dart
import 'dart:io';
import 'dart:typed_data';
import 'dart:convert';
import 'dart:math' as math;

import 'package:image/image.dart';
import 'package:mcd_dart/mcd_dart.dart';
import 'package:path/path.dart' as path;

import '../models/image.dart';
// Import when needed
// import '../models/card.dart';
import '../image/processing.dart';
import '../utils/image_hash.dart';
import '../utils/config.dart';
import 'recognition.dart';

class MagicCardDetector {
  String? outputPath;
  List<ReferenceImage> referenceImages = [];
  List<TestImage> testImages = [];
  
  bool verbose = false;
  bool visual = false;
  
  double hashSeparationThr = 4.0;
  int thrLvl = 70;
  
  MagicCardDetector({this.outputPath});
  
  Future<void> exportReferenceData(String path) async {
    // Export the hash and card name data for future use
    List<Map<String, dynamic>> hashData = [];
    
    for (var image in referenceImages) {
      if (image.phash != null) {
        hashData.add({
          'name': image.name,
          'hash': image.phash!.hashValue,
        });
      }
    }
    
    final jsonData = jsonEncode(hashData);
    await File(path).writeAsString(jsonData);
  }
  
  Future<void> readPrehashReferenceData(String path) async {
    print('Reading prehashed data from $path');
    print('...');
    
    try {
      final content = await File(path).readAsString();
      final List<dynamic> data = jsonDecode(content);
      
      for (var item in data) {
        final hash = ImageHash(item['hash'] as int);
        referenceImages.add(
          ReferenceImage(item['name'], null, phash: hash)
        );
      }
      
      print('Done. Loaded ${referenceImages.length} reference cards from $path.');
    } catch (e) {
      print('Error reading reference data: $e');
      rethrow;
    }
  }
  
  /// Loads all available set hash files from the assets/set_hashes directory
  Future<void> loadAllSetHashes() async {
    print('Loading all available set hashes...');
    
    try {
      final hashFiles = Config.getAvailableSetHashes();
      
      if (hashFiles.isEmpty) {
        print('No hash files found in ${Config.getSetHashesDirectory()}');
        return;
      }
      
      int totalCards = 0;
      for (var file in hashFiles) {
        final initialCount = referenceImages.length;
        await readPrehashReferenceData(file.path);
        final newCards = referenceImages.length - initialCount;
        totalCards += newCards;
        print('Loaded $newCards cards from ${path.basename(file.path)}');
      }
      
      print('Total cards loaded from all set hashes: $totalCards');
    } catch (e) {
      print('Error loading set hashes: $e');
      rethrow;
    }
  }
  
  Future<void> readAndAdjustReferenceImages(String directory) async {
    print('Reading images from $directory');
    print('...');
    
    final Directory dir = Directory(directory);
    final List<FileSystemEntity> entities = await dir.list().toList();
    final List<File> imageFiles = entities
        .whereType<File>()
        .where((File file) => file.path.toLowerCase().endsWith('.jpg'))
        .toList();
    
    for (File file in imageFiles) {
      final Uint8List bytes = await file.readAsBytes();
      final Image? img = decodeImage(bytes);
      
      if (img != null) {
        final String name = path.basename(file.path);
        referenceImages.add(ReferenceImage(name, img));
      }
    }
    
    print('Done. Loaded ${referenceImages.length} reference cards.');
  }
  
  Future<void> readAndAdjustTestImages(String directory) async {
    print('Reading images from $directory');
    print('...');
    
    final int maxSize = 1000;
    
    final Directory dir = Directory(directory);
    final List<FileSystemEntity> entities = await dir.list().toList();
    final List<File> imageFiles = entities
        .whereType<File>()
        .where((File file) => file.path.toLowerCase().endsWith('.jpg'))
        .toList();
    
    for (File file in imageFiles) {
      final Uint8List bytes = await file.readAsBytes();
      Image? img = decodeImage(bytes);
      
      if (img != null) {
        // Resize if needed
        if (math.min(img.width, img.height) > maxSize) {
          final double scalef = maxSize / math.min(img.width, img.height);
          img = copyResize(
            img,
            width: (img.width * scalef).toInt(),
            height: (img.height * scalef).toInt(),
            interpolation: Interpolation.average
          );
        }
        
        final String name = path.basename(file.path);
        testImages.add(TestImage(name, img));
      }
    }
    
    print('Done. Loaded ${testImages.length} test images.');
  }
  
  RecognitionResult recognizeSegment(Image imageSegment) {
    // Wrapper for different recognition algorithms
    return phashCompare(
      imageSegment, 
      referenceImages,
      hashSeparationThr: hashSeparationThr,
      verbose: verbose
    );
  }
  
  void recognizeCardsInImage(TestImage testImage, String contouringMode) {
    print('Segmenting card candidates out of the image...');
    print('Using $contouringMode algorithm.');
    
    testImage.candidateList.clear();
    segmentImage(testImage, contouringMode: contouringMode);
    
    print('Done. Found ${testImage.candidateList.length} candidates.');
    print('Recognizing candidates.');
    
    for (int iCand = 0; iCand < testImage.candidateList.length; iCand++) {
      CardCandidate candidate = testImage.candidateList[iCand];
      
      if (verbose) {
        print('${iCand + 1} / ${testImage.candidateList.length}');
      }
      
      // Easy fragment / duplicate detection
      bool isFragment = false;
      for (CardCandidate otherCandidate in testImage.candidateList) {
        if (otherCandidate.isRecognized && !otherCandidate.isFragment) {
          if (otherCandidate.contains(candidate)) {
            candidate.isFragment = true;
            isFragment = true;
            break;
          }
        }
      }
      
      if (!isFragment) {
        // Recognize the segment
        final RecognitionResult result = recognizeSegment(candidate.image);
        candidate.isRecognized = result.isRecognized;
        candidate.recognitionScore = result.recognitionScore;
        candidate.name = result.cardName;
      }
    }
    
    print('Done. Found ${testImage.returnRecognized().length} cards.');
    if (verbose) {
      for (CardCandidate card in testImage.returnRecognized()) {
        print('${card.name}; S = ${card.recognitionScore}');
      }
    }
    
    print('Removing duplicates...');
    // Final fragment detection
    testImage.markFragments();
    print('Done.');
  }
  
  Future<List<Uint8List>> processImage(Uint8List imageBytes, String imageName) async {
    print('\n--- Processing Image: $imageName ---');
    
    // Decode the image
    final Image? image = decodeImage(imageBytes);
    if (image == null) {
      throw Exception('Failed to decode image');
    }
    
    // Create a temporary TestImage
    final TestImage testImage = TestImage(imageName, image);
    
    // Try different algorithms
    final List<String> algList = ['adaptive', 'rgb'];
    
    for (String alg in algList) {
      recognizeCardsInImage(testImage, alg);
      testImage.discardUnrecognizedCandidates();
      
      // Stop if we've found enough cards or exhausted search potential
      if (!testImage.mayContainMoreCards() || 
          testImage.returnRecognized().length > 5) {
        break;
      }
    }
    
    // Generate result images
    print('Generating result images...');
    
    // Original image bytes
    final Uint8List originalBytes = encodeJpg(testImage.original);
    
    // Annotated image
    final Image annotatedImage = testImage.plotImageWithRecognized();
    final Uint8List annotatedBytes = encodeJpg(annotatedImage);
    
    print('Done.');
    
    // Return both images
    return [originalBytes, annotatedBytes];
  }
  
  Future<void> runRecognition([List<int>? imageIndices]) async {
    // The top-level image recognition method
    final List<int> indices = imageIndices ?? List<int>.generate(testImages.length, (int i) => i);
    
    for (int i in indices) {
      final TestImage testImage = testImages[i];
      print('Accessing image ${testImage.name}');
      
      if (visual) {
        print('Original image displayed');
      }
      
      final List<String> algList = ['adaptive', 'rgb'];
      
      for (String alg in algList) {
        recognizeCardsInImage(testImage, alg);
        testImage.discardUnrecognizedCandidates();
        
        if (!testImage.mayContainMoreCards() || 
            testImage.returnRecognized().length > 5) {
          break;
        }
      }
      
      print('Plotting and saving the results...');
      
      // Create annotated image
      final Image resultImage = testImage.plotImageWithRecognized();
      
      // Save to file if output path is provided
      if (outputPath != null) {
        final String fileName = testImage.name.contains('.jpg') 
            ? testImage.name.split('.jpg')[0] 
            : testImage.name;
        
        final String outFileName = path.join(
          outputPath!, 
          'MTG_card_recognition_results_$fileName.jpg'
        );
        
        final File outFile = File(outFileName);
        await outFile.writeAsBytes(encodeJpg(resultImage));
      }
      
      print('Done.');
      
      // Print recognized cards
      final List<CardCandidate> recognizedList = testImage.returnRecognized();
      print('Recognized cards (${recognizedList.length} cards):');
      for (CardCandidate card in recognizedList) {
        print('${card.name} - with score ${card.recognitionScore}');
      }
    }
    
    print('Recognition done.');
  }
}